Prague updates

fselmo opened the pull request:
### What was wrong?

We need to add some additional support for Prague-related fields.

### How was it fixed?

- Add formatters for authorization list fields and block requests field in block headers
- Add integration test for 7702 so both geth and py-evm interactions are tested with expected results

--- 
Significant change to integration tests:

Change integration tests so that each test has its own context. We are having way too many issues with muddied contexts and tests needing to wait on the right moment to test transactions between a mining state. This isolates each integration test into its own state which makes issues a lot easier to track and should give less CI headaches.

---

### Todo:

- [x] Clean up commit history
- [x] Add or update documentation related to these changes --> This probably needs to be better... future PR + blog post?
- [x] Add at least one good 7702 tx test to the integration tests so we can test across both geth and eth-tester <-> py-evm
- [x] Will need ethereum/eth-tester#317 and a release which is waiting on a *py-evm* Prague release. Update eth-tester lower pin when release is ready.
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

<img width="714" alt="Screenshot 2025-04-23 at 13 19 00" src="https://github.com/user-attachments/assets/1ec4642e-289f-4a19-8e51-4905f76014d3" />

fselmo committed:
Add formatter support for Prague header field requests_hash

fselmo committed:
Add formatters for set code transaction support

- Add formatters for authorization lists
- Provide instructions for serializing pydantic
  ``BaseModel``. In the future we may want to tighten
  this up to expect the ``CamelModel`` introduced in
  `eth-utils` ``5.3.0``.

fselmo committed:
remove unused vars

fselmo committed:
Add tests for set code transactions

- Add to the required formatter changes for eth-tester support
- Add integration test for set code tx
- Add set code core test to test `send_transaction` via eth-tester

fselmo committed:
Add v8 TODO; bump eth-account lower pin for 7702 support

fselmo committed:
bump eth-tester to newly released version with prague changes

- newsfragment for #3659

fselmo committed:
Test a more interesting test case for 7702:

- Change the storage of the EOA in the same transaction by
  delegating to the ``math_contract`` and calling increment
  counter on itself.

fselmo committed:
Change pytest scope to function for all integration tests:

- Instantiate a new geth instance for each test to avoid muddied
  context between tests

- Lower the ``dev.period`` to ``1`` since the tests don't need to account for
  each other anymore

- Parallelize the tests in CI with pytest-xdist (``-n 10``)

- Properly disconnect the ``async_await_w3`` ws tests

fselmo committed:
Let geth choose the open port, pass it to fixtures

- Let geth choose the open port and parse this port from the logs
  to pass along to the consuming fixtures.

fselmo committed:
Some cleanup on integration test refactor; skip port for ipc

fselmo commented:
I fixed the integration test issues 🙂  I think they are better this way... every test runs in its own context now and we can parallelize them with xdist in the CI which doesn't seem to cause any issues on the `xlarge` setup with `-n 10`.

This is ready for a review 👀

fselmo committed:
Don't check blocknum equality, we could be at block transition

fselmo committed:
Update contributing docs for parallelizing integration tests

kclowes commented on docs/contributing.rst:
```diff
@@ -200,6 +201,16 @@ tests are written across different files within ``/web3/_utils/module_testing``.
+
+.. code:: sh
+
+    $ pytest tests/integration/go_ethereum/path/to/module/or/test -n 4
```
maybe:
```suggestion
    $ pytest tests/integration/go_ethereum/path/to/module/or/test -n auto
```

kclowes commented on tox.ini:
```diff
@@ -20,17 +20,18 @@ allowlist_externals=make,pre-commit
+    integration-goethereum-http_async: pytest {posargs:tests/integration/go_ethereum/test_goethereum_http.py -k Async -n 10}
+    integration-goethereum-legacy_ws: pytest {posargs:tests/integration/go_ethereum/test_goethereum_legacy_ws.py -n 10}
+    integration-goethereum-ws: pytest {posargs:tests/integration/go_ethereum/test_goethereum_ws -n 10}
+    integration-ethtester: pytest {posargs:tests/integration/test_ethereum_tester.py -n 10}
```
I wonder if auto would be better here? then we're not hard coding a certain number.

fselmo replied:
My only hesitation was that `auto` could end up being _more_ than `10`. I was sort of trying to set some lower limit. But I guess if it can handle more than `10` that would be ok?

fselmo replied:
Thoughts on that?

fselmo replied:
I added it, let's see what auto does for the `xlarge` CI runs

kclowes commented on tox.ini:
```diff
@@ -20,17 +20,18 @@ allowlist_externals=make,pre-commit
 install_command=python -m pip install {opts} {packages}
 usedevelop=True
 commands=
+    # Note: `pytest-xdist` n=10 seems to cause connection issues with geth; keep it at 6 for now
```
I think this can go?

kclowes commented on web3/_utils/method_formatters.py:
```diff
@@ -161,6 +164,12 @@ def type_aware_apply_formatters_to_dict(
     """
     Preserve ``AttributeDict`` types if original ``value`` was an ``AttributeDict``.
     """
+    # TODO: In v8, Use eth-utils 5.3.0 as lower pin where ``apply_formatters_to_dict``
```
Do we want to pull CamelModel into web3 in this PR?

fselmo replied:
I suppose we can. Should I add it to this PR?

kclowes replied:
I'll leave that up to you!

fselmo replied:
I added it in here 👍🏼

fselmo replied:
Oh we can't. `eth-account` can't inherit from ``CamelModel`` so we can't support it here yet since ``SignedSetCodeAuthorization`` is coming from eth-account.

Putting it back.

kclowes commented on web3/_utils/module_testing/eth_module.py:
```diff
@@ -3485,8 +3527,8 @@ def test_eth_replace_transaction_legacy(
         assert replace_txn["value"] == 1
         assert replace_txn["gas"] == 21000
         assert replace_txn["gasPrice"] == txn_params["gasPrice"]
+        print("yeeeeeeehhhboiiiiiiiii")
```
😆

fselmo replied:
Can you tell how much joy this gave me? 😆

kclowes replied:
lol yes I can!

kclowes reviewed (approved):
:shipit: Nothing major stood out to me. For my own learning, is the fixture scope from `module` -> `function` for ease of parallelization, or is there another reason? It looks like test times are down with the parallelization you added, but could we get more speedup if we scope the fixtures wider?

fselmo committed:
Changes from comments on PR #3659

fselmo commented:
> For my own learning, is the fixture scope from `module` -> `function` for ease of parallelization, or is there another reason? 

I think there are maybe a few fixtures we can change back to `module` but this is so that each test has its own instance of geth. This is the change that makes them run in isolation as each function is a single test.

---
edit: Even still, the overhead of _not_ changing only one or two fixtures to `module` or `session` is probably pretty low

fselmo committed:
Add upper bound on number of processes to run for CI runs as this can
  help control flakiness. ``36`` workers (unbounded ``-n auto``) seems
  to be too much to account for tests that rely on the ``dev.period``
  block processing time.

fselmo commented:
@kclowes all changes from your comments accounted for. I don't think the ``dev.period`` is going to affect test times other than the ones that get a / wait for a transaction receipt, which should be 1 or 2 tests. Aside from those, tests should finish as soon as the test logic passes the test, which for most tests doesn't require block processing. I raised the ``dev.period`` to ``4`` in this PR since ``-n auto`` increased the number of workers to ~ ``36`` and slowed down processing time for each test, which ended up causing some issues due to mining.

In the end ``-n auto`` with increased ``dev.period`` seems like the right way to go. So far, running the CI with ``--dev.period=4`` seems to be plenty of time for the pending / replace transaction tests to run their logic before the block is processed.

I'm going to re-run the CI a few more times and make sure the success rate good and then I'll merge this.

reedsa reviewed (approved):
👍

pacrob reviewed (approved):
:sheep:

fselmo committed:
Silence warnings on async iterator by properly disconnecting
